racket/collects/racket/list.rkt

```diff
@@ -129,16 +131,61 @@
   (unless (and (procedure? f)
                (procedure-arity-includes? f 1))
     (raise-argument-error 'list-update "(-> any/c any/c)" 2 l i f))
-  (cond
-   [(zero? i) (cons (f (car l)) (cdr l))]
-   [else (cons (car l) (list-update (cdr l) (sub1 i) f))]))
+  (let loop ([l l] [i i])
+    (cond
+      [(zero? i) (cons (f (car l)) (cdr l))]
+      [else (cons (car l) (loop (cdr l) (sub1 i)))])))
+
+(define (list-update* l . pairs)
+  (unless (list? l)
+    (apply raise-argument-error 'list-update* "list?" 0 l pairs))
+  (unless (even? (length pairs))
+    (error 'list-update* "expected an even number of association elements, but received an odd number: ~e" pairs))
```

Comment:
IIUC, we should avoid using `error`, since it does not follow the error message conventions.

Here's how this kind of pairing errors are reported in `racket/collects/racket/private/hash.rkt`'s `paired-fold`:

```
(raise-arguments-error
          who
          (format "expected ~a, but received ~a"
                  "an even number of association elements"
                  "an odd number of association elements")
          "association elements"
          pairs0)
```